reference-implementation/lib/writable-stream.js

```diff
@@ -4,225 +4,661 @@ const { InvokeOrNoop, PromiseInvokeOrNoop, PromiseInvokeOrFallbackOrNoop, Valida
+      return Promise.reject(new TypeError('cannot close an already-closing stream'));
     }
 
-    assert(this._state === 'waiting' || this._state === 'writable');
+    return WritableStreamDefaultWriterClose(this);
+  }
 
-    let chunkSize = 1;
+  releaseLock() {
+    if (IsWritableStreamDefaultWriter(this) === false) {
+      throw CreateWritableStreamDefaultWriterBrandCheckException('releaseLock');
+    }
 
-    if (this._strategySize !== undefined) {
-      try {
-        chunkSize = this._strategySize(chunk);
-      } catch (chunkSizeE) {
-        ErrorWritableStream(this, chunkSizeE);
-        return Promise.reject(chunkSizeE);
-      }
+    const stream = this._ownerWritableStream;
+
+    if (stream === undefined) {
+      return undefined;
     }
 
-    let resolver, rejecter;
-    const promise = new Promise((resolve, reject) => {
-      resolver = resolve;
-      rejecter = reject;
-    });
+    assert(stream._writer !== undefined);
 
-    const writeRecord = { promise: promise, chunk: chunk, _resolve: resolver, _reject: rejecter };
-    try {
-      EnqueueValueWithSize(this._queue, writeRecord, chunkSize);
-    } catch (enqueueResultE) {
-      ErrorWritableStream(this, enqueueResultE);
-      return Promise.reject(enqueueResultE);
+    const state = stream._state;
+
+    const releasedException = new TypeError('Writer was released and can no longer be used to monitor the stream\'s closedness');
+
+    if (state === 'writable' || state === 'waiting' || state === 'closing') {
+      WritableStreamDefaultWriterClosedPromiseReject(this, releasedException);
+    } else {
+      WritableStreamDefaultWriterClosedPromiseResetToRejected(this, releasedException);
     }
 
-    SyncWritableStreamStateWithQueue(this);
-    CallOrScheduleWritableStreamAdvanceQueue(this);
-    return promise;
+    if (state === 'waiting') {
+      WritableStreamDefaultWriterReadyPromiseReject(this, releasedException);
+    } else {
+      WritableStreamDefaultWriterReadyPromiseResetToRejected(this, releasedException);
+    }
+
+    stream._writer = undefined;
+    this._ownerWritableStream = undefined;
+  }
+
+  write(chunk) {
+    if (IsWritableStreamDefaultWriter(this) === false) {
+      return Promise.reject(CreateWritableStreamDefaultWriterBrandCheckException('write'));
+    }
+
+    if (this._ownerWritableStream === undefined) {
+      return Promise.reject(CreateWritableStreamDefaultWriterLockException('write to'));
+    }
+
+    if (this._ownerWritableStream._state === 'closing') {
+      return Promise.reject(new TypeError('Cannot write to an already-closed stream'));
+    }
+
+    return WritableStreamDefaultWriterWrite(this, chunk);
   }
 }
 
-exports.WritableStream = WritableStream;
+// Helper functions for the WritableStreamDefaultWriter.
 
-function closure_WritableStreamErrorFunction() {
-  const f = e => ErrorWritableStream(f._stream, e);
-  return f;
+function CreateWritableStreamDefaultWriterBrandCheckException(name) {
```

Comment:
Name this lowercase since it isn't an abstract op. Also since it's local to this file we could just call it `writerBrandCheckException`.

(Similar comments for the other helpers. Maybe also consider moving the helpers to one section instead of per class?)